include/mqtt/property_parse.hpp

```diff
@@ -17,273 +17,229 @@ namespace mqtt {
 namespace v5 {
 namespace property {
 
-template <typename It>
 inline
-mqtt::optional<property_variant> parse_one(It& begin, It end) {
-    if (begin == end) return mqtt::nullopt;
+std::pair<mqtt::optional<property_variant>, buffer> parse_one(buffer buf) {
```

Comment:
I don't see why this function should return an std::pair.

The properties, if any are successfully parsed, will automatically preserve the lifetime of the mqtt::buffer object.

No need to explicitly return the buffer object.

Further, I don't think this function should have that responsibility.  Callers of it should be aware that if no properties are returned, then there may be no references to `buf`. So they would need to ensure they manage the lifetime the way they want to.

Also, for functions like this one where the number of "assignments" of the parameter is unknown, I think it should be a const& parameter, instead of by-value.

Reply:
The reason I return `buf` as the second member of the `pair` is updating the `buffer` position to the next property starting point.

We can pass an iterator with buf. See the following code:

```cpp
inline
std::pair<std::vector<property_variant>, buffer> parse(buffer buf) {
    std::vector<property_variant> props;
    auto it = buf.cbegin();
    while (true) {
        auto ret = parse_one(buf, it); // buf is passed by const reference, 
                                       // it is passed by non const reference 
        // it is updated to the next property start position
        if (ret.first) {
            props.push_back(std::move(ret.first.value()));
            buf = std::move(ret.second);
        }
        else {
            break;
        }
    }
    return { props, std::move(buf) };
}
```

Or we can also use https://github.com/redboltz/mqtt_cpp/pull/337#discussion_r313728244 approach.

In addition `parse` doesn't need to return `buffer`. The caller knows whole properties size by Property Length field.

Any ideas?

Reply:
The property length field is always exactly the same, regardless of which type of property.

I would parse the property length outside of the parse_one function, and then create an MQTT::string_view with the whole data for the property. Then pass the MQTT::string_view by value, and but by const&.

If the parse_one function returns a property that needs to own a reference, then it will make a new MQTT::buffer using the lifetime from the const& buf, and the data from the MQTT::string_view.

Reply:
Property length field doesn't appear each properties. 

I mean the data structure is as follows:

```
+-----------------+-----------+-----------+-----+-----------+
| Prioerty Length | Property1 | Property2 | ... | PropertyN |
+-----------------+-----------+-----------+-----+-----------+
```

We cannot know each property size and number of properties.

See https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901027